core/cloudflow-spark/src/main/scala/cloudflow/spark/SparkStreamlet.scala

```diff
@@ -84,59 +79,51 @@ trait SparkStreamlet extends Streamlet[SparkStreamletContext] with Serializable
   protected def createLogic(): SparkStreamletLogic
 
   override final def run(context: SparkStreamletContext): StreamletExecution = {
+    val completionPromise = Promise[Dun]()
+    val completionFuture  = completionPromise.future
+
     val InitialDelay                 = 2 seconds
     val MonitorFrequency             = 5 seconds
     implicit val system: ActorSystem = ActorSystem("spark_streamlet", context.config)
 
-    readyPromise.trySuccess(Dun)
     val streamletQueryExecution = createLogic.buildStreamingQueries
 
     new StreamletExecution {
-      val readyFuture = readyPromise.future
-
-      // schedule a function to check periodically if any of the queries
-      // raised an exception
-      val done = system.scheduler.schedule(InitialDelay, MonitorFrequency) {
-        val failed = streamletQueryExecution.queries.filter(_.exception.nonEmpty)
-        if (failed.nonEmpty) {
-          // if any of the queries has an exception, stop them all
-          val errors = failed.map(f ⇒ f.name -> f.status.message).mkString(", ")
-          system.log.error(s"Queries failed. Stopping Process. Reason: $errors")
 
-          // stop all query execution
-          streamletQueryExecution.stop()
-
-        } else ()
+      // schedule a function to check periodically if any of the queries stopped
+      val scheduledQueryCheck: Cancellable = system.scheduler.schedule(InitialDelay, MonitorFrequency) {
+        val someQueryStopped = streamletQueryExecution.queries.exists(!_.isActive)
+        if (someQueryStopped) completionPromise.completeWith(stop())
       }
 
       // this future will be successful when any of the queries face an exception
       // or is stopped. The runner needs to await on this future and exit only when it
       // succeeds
-      def completed: Future[Dun] = {
-        if (streamletQueryExecution.queries.forall(!_.isActive)) {
-          // not active means the queries have either been stopped or they faced an exception
-          done.cancel()
-
+      def completed: Future[Dun] = completionFuture
+      val ready: Future[Dun]     = Future.successful(Dun)
+      def stop(): Future[Dun] = {
+        streamletQueryExecution.stop()
+        scheduledQueryCheck.cancel()
+        Future {
+          val t0 = System.currentTimeMillis()
+          // Step 1: Wait for all queries to stop
+          while (streamletQueryExecution.queries.exists(_.isActive)) {
+            val elapsedTime = System.currentTimeMillis() - t0
+            if (elapsedTime < StopTimeout) {
+              Thread.sleep(1.second.toMillis)
+            } else {
+              throw new QueryTerminationTimeout(streamletQueryExecution.queries.filter(_.isActive), elapsedTime)
+            }
```

Comment:
Can we replace this with something like https://gist.github.com/viktorklang/9414163 .. implement a retry with backoff semantics ?

Reply:
It's on `stop` so it's not the most critical path, but yes, I agree it can be polished a bit and use some async-fu ;-)